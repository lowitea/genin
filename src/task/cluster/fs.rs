use std::{
    error::Error as StdError,
    fmt::{self, Display},
    fs::File,
    io::{self, Read, Write},
    path::PathBuf,
};

use clap::ArgMatches;
use log::{debug, warn};
use serde::{de::DeserializeOwned, Serialize};

use crate::{
    error::{GeninError, GeninErrorKind},
    task::{serde_genin, Validate},
};

pub const CLUSTER_YAML: &str = "cluster.genin.yml";
pub const INVENTORY_YAML: &str = "inventory.yml";
pub const UPGRADE_YAML: &str = "upgrade.genin.yml";

//TODO: remove it in next commits
#[allow(unused)]
#[derive(Default)]
pub(in crate::task) struct FsInteraction {
    source: Option<PathBuf>,
    output: Option<PathBuf>,
}

impl<'a> From<&'a ArgMatches> for FsInteraction {
    fn from(args: &'a ArgMatches) -> Self {
        FsInteraction {
            source: args
                .try_get_one::<&str>("source")
                .unwrap_or_default()
                .map(PathBuf::from),
            output: args
                .try_get_one::<&str>("output")
                .unwrap_or_default()
                .map(PathBuf::from),
        }
    }
}

//TODO: remove in future commits
#[allow(unused)]
impl FsInteraction {
    /// Reading source file
    ///
    /// # panic
    /// - source file does not exist
    /// - source file wrong format
    pub fn read(&self) -> Result<Vec<u8>, GeninError> {
        let mut file = File::open(self.source.as_ref().ok_or_else(|| {
            GeninError::new(
                GeninErrorKind::ArgsError,
                "Error while trying to read source file. Source file: None",
            )
        })?)
        .map_err(|error| {
            GeninError::new(
                GeninErrorKind::ArgsError,
                format!(
                    "Error then opening file {}! Err: {}",
                    self.source.as_ref().unwrap().to_str().unwrap(),
                    error
                )
                .as_str(),
            )
        })?;
        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer).map_err(|error| {
            GeninError::new(
                GeninErrorKind::ArgsError,
                format!(
                    "Error then opening file {}! Err: {}",
                    self.source.as_ref().unwrap().to_str().unwrap(),
                    error
                )
                .as_str(),
            )
        })?;
        Ok(buffer)
    }
}

#[inline]
/// Check that file with current extension exists and replace path if it
/// exists with related extension.
fn right_ext(path: PathBuf, path_str: String, second_try: bool) -> Option<PathBuf> {
    match (
        path.is_file(),
        path.extension().and_then(|e| e.to_str()),
        second_try,
    ) {
        (false, Some("yml"), false) => {
            let new_path_str = path_str.replace(".yml", ".yaml");
            debug!("file_exists: false, extension: .yml, second_try: false");
            warn!(
                "file {} does not exists, trying to open {}",
                path_str, &new_path_str
            );
            right_ext(PathBuf::from(&new_path_str), new_path_str, true)
        }
        (false, Some("yaml"), false) => {
            let new_path_str = path_str.replace(".yaml", ".yml");
            debug!("file_exists: false, extension: .yaml, second_try: false");
            warn!(
                "file {} does not exists, trying to open {}",
                path_str, &new_path_str
            );
            right_ext(PathBuf::from(&new_path_str), new_path_str, true)
        }
        (false, _, true) => None,
        _ => Some(path),
    }
}

pub trait TryIntoFile {
    type Error;

    fn try_into_file(self) -> Result<Option<File>, Self::Error>;
}

impl TryIntoFile for Option<PathBuf> {
    type Error = std::io::Error;

    fn try_into_file(self) -> Result<Option<File>, Self::Error> {
        if let Some(path) = self {
            return File::open(path).map(Some);
        }

        Ok(None)
    }
}

pub struct IO<I, O> {
    pub input: Option<I>,
    pub output: Option<O>,
}

#[allow(unused)]
impl IO<(), ()> {
    pub fn new() -> Self {
        Self {
            input: None,
            output: None,
        }
    }
}

impl Default for IO<(), ()> {
    fn default() -> Self {
        Self {
            input: Default::default(),
            output: Default::default(),
        }
    }
}

impl<'a> From<&'a ArgMatches> for IO<PathBuf, PathBuf> {
    fn from(args: &'a ArgMatches) -> Self {
        Self {
            input: args
                .try_get_one::<String>("source")
                .transpose()
                .and_then(|r| r.map_or(None, |s| Some(PathBuf::from(s.as_str())))),
            output: args
                .try_get_one::<String>("output")
                .transpose()
                .and_then(|r| r.map_or(None, |s| Some(PathBuf::from(s.as_str())))),
        }
    }
}

impl IO<PathBuf, PathBuf> {
    pub fn try_into_files(
        self,
        source: Option<&str>,
        output: Option<&str>,
        force: bool,
    ) -> Result<IO<File, File>, Box<dyn StdError>> {
        Ok(IO {
            input: self
                .input
                .or_else(|| source.map(PathBuf::from))
                .and_then(|path| {
                    path.clone()
                        .to_str()
                        .map(|path_str| (path, path_str.to_string()))
                })
                .and_then(|(path, path_str)| right_ext(path, path_str, false))
                .map(File::open)
                .transpose()?,
            output: self
                // can be
                .output
                .or_else(|| output.map(PathBuf::from))
                .map(|path| {
                    if !path.exists() || force {
                        File::create(path)
                    } else if path.is_file() {
                        let file_ext = path
                            .extension()
                            .expect("Failed to get file extension")
                            .to_str()
                            .expect("Failed to cast into str");
                        let file_name = path
                            .file_stem()
                            .expect("Failed to get filename")
                            .to_str()
                            .expect("Failed to cast into str");

                        warn!(
                            "the target file {} already exists so the new file will be \
                                saved with name {file_name}.copy.{file_ext}",
                            path.display(),
                        );

                        File::create(format!("{file_name}.copy.{file_ext}"))
                    } else {
                        Err(io::Error::new(
                            io::ErrorKind::Other,
                            "output is not file or not valid path",
                        ))
                    }
                })
                .transpose()?,
        })
    }
}

impl<I: Read, O> IO<I, O> {
    pub fn deserialize_input<T>(self) -> Result<IO<T, O>, Box<dyn StdError>>
    where
        T: DeserializeOwned + fmt::Debug + Validate + 'static,
    {
        let mut bytes = Vec::new();
        self.input
            .ok_or_else(|| {
                GeninError::new(
                    GeninErrorKind::EmptyField,
                    "IO struct has empty input field. Maybe the file doesn't exist!",
                )
            })?
            .read_to_end(&mut bytes)?;
        Ok(IO {
            input: Some(serde_genin::from_slice(&bytes)?),
            output: self.output,
        })
    }
}

pub trait TryMap<A, B> {
    type Error;
    type Output;

    fn try_map<F>(self, f: F) -> Result<Self::Output, Self::Error>
    where
        Self: Sized,
        F: FnOnce(Self) -> Result<Self::Output, Self::Error>;
}

impl<I, O, A, B> TryMap<A, B> for IO<I, O> {
    type Error = GeninError;
    type Output = IO<A, B>;

    fn try_map<F>(self, function: F) -> Result<Self::Output, Self::Error>
    where
        Self: Sized,
        F: FnOnce(Self) -> Result<Self::Output, Self::Error>,
    {
        function(self)
    }
}

impl<I: Display, O> IO<I, O> {
    pub fn print_input(self, quiet: bool) -> Self {
        if let (Some(input), false) = (self.input.as_ref(), quiet) {
            println!("{input}")
        }
        self
    }
}

impl<I: Serialize, O: Write> IO<I, O> {
    pub fn serialize_input(self) -> Result<IO<I, O>, Box<dyn StdError>> {
        if let IO {
            input,
            output: Some(mut writer),
        } = self
        {
            serde_yaml::to_writer(&mut writer, &input)?;
            Ok(IO {
                input,
                output: Some(writer),
            })
        } else {
            Err(GeninError::new(
                GeninErrorKind::Serialization,
                "failed to serialize input because output file is None",
            )
            .into())
        }
    }
}

impl<I, O> IO<I, O> {
    pub fn consume_output(self) -> IO<I, String> {
        IO {
            input: self.input,
            output: Option::<String>::None,
        }
    }
}

impl<I, O> Display for IO<I, O>
where
    I: Display,
    O: Display,
{
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            IO {
                input: Some(input),
                output: Some(output),
            } => write!(f, "{}{}", input, output),
            IO {
                input: Some(input),
                output: None,
            } => write!(f, "{}", input),
            IO {
                input: None,
                output: Some(output),
            } => write!(f, "{}", output),
            _ => write!(f, ""), //TODO
        }
    }
}

//#[cfg(test)]
//mod test;
