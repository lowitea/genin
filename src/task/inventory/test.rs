use std::convert::TryFrom;

use crate::task::cluster::Cluster;

use super::Inventory;

#[test]
fn inventory_from_cluster_v1() {
    let cluster_v1_str: String = r#"---
instances:
  - name: api
    count: 1
    replicas: 0
    weight: 10
    roles:
      - api
  - name: calculator
    count: 1
    replicas: 0
    weight: 10
    roles:
      - calculator
  - name: storage
    type: storage
    count: 2
    replicas: 2
    weight: 10
    roles:
      - storage
    config:
      memtx_memory: 1179869184
      vinyl_memory: 10345452343
  - name: cache
    type: storage
    count: 2
    replicas: 2
    weight: 10
    roles:
      - storage
  - name: router
    type: router
    count: 1
    replicas: 0
    weight: 10
    roles:
      - router
      - failover-coordinator
hosts:
  - name: docker
    type: datacenter
    ports:
      http: 8100
      binary: 5002
    hosts:
      - name: host-1
        ip: 10.99.16.65
      - name: host-2
        ip: 10.99.16.66
failover:
  mode: stateful
  state_provider: stateboard
  stateboard_params:
    uri:
      ip: 10.99.16.66
      port: 5001
    password: genin-app
vars:
  ansible_user: ansible
  ansible_password: ansible
  cartridge_app_name: genin-app
  cartridge_cluster_cookie: genin-app-secret-cookie"#
        .into();

    let inventory_model_str: String = r#"---
all:
  vars:
    ansible_user: ansible
    ansible_password: ansible
    cartridge_app_name: genin-app
    cartridge_cluster_cookie: genin-app-secret-cookie
    cartridge_failover_params:
      mode: stateful
      state_provider: stateboard
      stateboard_params:
        uri: "10.99.16.66:5001"
        password: genin-app
  hosts:
    api-1:
      config:
        advertise_uri: "10.99.16.65:5002"
        http_port: 8100
    storage-1-1:
      config:
        advertise_uri: "10.99.16.65:5003"
        http_port: 8101
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    storage-1-3:
      config:
        advertise_uri: "10.99.16.65:5004"
        http_port: 8102
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    storage-2-2:
      config:
        advertise_uri: "10.99.16.65:5005"
        http_port: 8103
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    cache-1-1:
      config:
        advertise_uri: "10.99.16.65:5006"
        http_port: 8104
    cache-1-3:
      config:
        advertise_uri: "10.99.16.65:5007"
        http_port: 8105
    cache-2-2:
      config:
        advertise_uri: "10.99.16.65:5008"
        http_port: 8106
    router-1:
      config:
        advertise_uri: "10.99.16.65:5009"
        http_port: 8107
    calculator-1:
      config:
        advertise_uri: "10.99.16.66:5002"
        http_port: 8100
    storage-1-2:
      config:
        advertise_uri: "10.99.16.66:5003"
        http_port: 8101
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    storage-2-1:
      config:
        advertise_uri: "10.99.16.66:5004"
        http_port: 8102
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    storage-2-3:
      config:
        advertise_uri: "10.99.16.66:5005"
        http_port: 8103
        memtx_memory: 1179869184
        vinyl_memory: 10345452343
    cache-1-2:
      config:
        advertise_uri: "10.99.16.66:5006"
        http_port: 8104
    cache-2-1:
      config:
        advertise_uri: "10.99.16.66:5007"
        http_port: 8105
    cache-2-3:
      config:
        advertise_uri: "10.99.16.66:5008"
        http_port: 8106
    stateboard:
      stateboard: true
      config:
        listen: "10.99.16.66:5001"
        password: genin-app
  children:
    api-1-replicaset:
      vars:
        replicaset_alias: api-1
        failover_priority:
          - api-1
        roles:
          - api
        weight: 10
      hosts:
        api-1: ~
    storage-1-replicaset:
      vars:
        replicaset_alias: storage-1
        failover_priority:
          - storage-1-1
          - storage-1-2
          - storage-1-3
        roles:
          - storage
        weight: 10
      hosts:
        storage-1-1: ~
        storage-1-3: ~
        storage-1-2: ~
    storage-2-replicaset:
      vars:
        replicaset_alias: storage-2
        failover_priority:
          - storage-2-1
          - storage-2-2
          - storage-2-3
        roles:
          - storage
        weight: 10
      hosts:
        storage-2-2: ~
        storage-2-1: ~
        storage-2-3: ~
    cache-1-replicaset:
      vars:
        replicaset_alias: cache-1
        failover_priority:
          - cache-1-1
          - cache-1-2
          - cache-1-3
        roles:
          - storage
        weight: 10
      hosts:
        cache-1-1: ~
        cache-1-3: ~
        cache-1-2: ~
    cache-2-replicaset:
      vars:
        replicaset_alias: cache-2
        failover_priority:
          - cache-2-1
          - cache-2-2
          - cache-2-3
        roles:
          - storage
        weight: 10
      hosts:
        cache-2-2: ~
        cache-2-1: ~
        cache-2-3: ~
    router-1-replicaset:
      vars:
        replicaset_alias: router-1
        failover_priority:
          - router-1
        roles:
          - router
          - failover-coordinator
        weight: 10
      hosts:
        router-1: ~
    calculator-1-replicaset:
      vars:
        replicaset_alias: calculator-1
        failover_priority:
          - calculator-1
        roles:
          - calculator
        weight: 10
      hosts:
        calculator-1: ~
    host-1:
      vars:
        ansible_host: 10.99.16.65
      hosts:
        api-1: ~
        storage-1-1: ~
        storage-1-3: ~
        storage-2-2: ~
        cache-1-1: ~
        cache-1-3: ~
        cache-2-2: ~
        router-1: ~
    host-2:
      vars:
        ansible_host: 10.99.16.66
      hosts:
        calculator-1: ~
        storage-1-2: ~
        storage-2-1: ~
        storage-2-3: ~
        cache-1-2: ~
        cache-2-1: ~
        cache-2-3: ~
        stateboard: ~
"#
    .into();

    let cluster: Cluster = serde_yaml::from_str(&cluster_v1_str).unwrap();

    let inventory = Inventory::try_from(&Some(cluster)).unwrap();

    let inventory_str = serde_yaml::to_string(&inventory).unwrap();

    assert_eq!(inventory_str, inventory_model_str);
}

#[test]
fn inventory_per_instance_vars() {
    let cluster_v2_str: String = r#"---
topology:
  - name: router
    replicasets_count: 1
    roles:
      - router
    vars:
      cartridge_extra_env:
        TARANTOOL_NET_MSG_MAX: 1536
      cartridge_force_leader_control_instance: false
hosts:
  - name: datacenter-1
    hosts:
      - name: server-1
        config:
          http_port: 8081
          binary_port: 3031
          address: 192.168.16.11
failover:
  mode: stateful
  state_provider: stateboard
  stateboard_params:
    uri: "192.168.16.11:4401"
    password: password
vars:
  ansible_user: ansible
  ansible_password: ansible
  cartridge_app_name: myapp
  cartridge_cluster_cookie: myapp-cookie
  cartridge_package_path: /tmp/myapp.rpm
  cartridge_bootstrap_vshard: true"#
        .into();

    let cluster = serde_yaml::from_str(&cluster_v2_str).unwrap();

    let inventory_model_str: String = r#"---
all:
  vars:
    ansible_user: ansible
    ansible_password: ansible
    cartridge_app_name: myapp
    cartridge_cluster_cookie: myapp-cookie
    cartridge_package_path: /tmp/myapp.rpm
    cartridge_bootstrap_vshard: true
    cartridge_failover_params:
      mode: stateful
      state_provider: stateboard
      stateboard_params:
        uri: "192.168.16.11:4401"
        password: password
  hosts:
    router-1:
      config:
        advertise_uri: "192.168.16.11:3031"
        http_port: 8081
      vars:
        cartridge_extra_env:
          TARANTOOL_NET_MSG_MAX: 1536
        cartridge_force_leader_control_instance: false
    stateboard:
      stateboard: true
      config:
        listen: "192.168.16.11:4401"
        password: password
  children:
    router-1-replicaset:
      vars:
        replicaset_alias: router-1
        failover_priority:
          - router-1
        roles:
          - router
      hosts:
        router-1: ~
    server-1:
      vars:
        ansible_host: 192.168.16.11
      hosts:
        router-1: ~
        stateboard: ~"#
        .into();

    let inventory_model = serde_yaml::from_str(&inventory_model_str).unwrap();

    let inventory = Inventory::try_from(&Some(cluster)).unwrap();

    assert_eq!(inventory, inventory_model);
}

#[test]
fn all_rw_only_in_replicaset() {
    let cluster_v2_str: String = r#"---
topology:
  - name: storage
    replicasets_count: 1
    roles:
      - storage
    all_rw: true
hosts:
  - name: datacenter-1
    hosts:
      - name: server-1
        config:
          http_port: 8081
          binary_port: 3031
          address: 192.168.16.11
failover:
  mode: stateful
  state_provider: stateboard
  stateboard_params:
    uri: "192.168.16.11:4401"
    password: password
vars:
  ansible_user: ansible
  ansible_password: ansible
  cartridge_app_name: myapp
  cartridge_cluster_cookie: myapp-cookie
  cartridge_package_path: /tmp/myapp.rpm
  cartridge_bootstrap_vshard: true"#
        .into();

    let cluster: Cluster = serde_yaml::from_str(&cluster_v2_str).unwrap();

    let inventory_model_str: String = r#"---
all:
  vars:
    ansible_user: ansible
    ansible_password: ansible
    cartridge_app_name: myapp
    cartridge_cluster_cookie: myapp-cookie
    cartridge_package_path: /tmp/myapp.rpm
    cartridge_bootstrap_vshard: true
    cartridge_failover_params:
      mode: stateful
      state_provider: stateboard
      stateboard_params:
        uri: "192.168.16.11:4401"
        password: password
  hosts:
    storage-1:
      config:
        advertise_uri: "192.168.16.11:3031"
        http_port: 8081
    stateboard:
      stateboard: true
      config:
        listen: "192.168.16.11:4401"
        password: password
  children:
    storage-1-replicaset:
      vars:
        replicaset_alias: storage-1
        failover_priority:
          - storage-1
        roles:
          - storage
        all_rw: true
      hosts:
        storage-1: ~
    server-1:
      vars:
        ansible_host: 192.168.16.11
      hosts:
        storage-1: ~
        stateboard: ~"#
        .into();

    let inventory_model: Inventory = serde_yaml::from_str(&inventory_model_str).unwrap();

    let inventory = Inventory::try_from(&Some(cluster)).unwrap();

    assert_eq!(inventory, inventory_model);
}

#[test]
fn test_build_without_ansible_password() {
    let cluster_v2_str: String = r#"---
topology:
  - name: storage
    replicasets_count: 1
    roles:
      - storage
    all_rw: true
hosts:
  - name: datacenter-1
    hosts:
      - name: server-1
        config:
          http_port: 8081
          binary_port: 3031
          address: 192.168.16.11
failover:
  mode: stateful
  state_provider: stateboard
  stateboard_params:
    uri: "192.168.16.11:4401"
    password: password
vars:
  cartridge_app_name: myapp
  cartridge_cluster_cookie: myapp-cookie
  cartridge_package_path: /tmp/myapp.rpm
  cartridge_bootstrap_vshard: true"#
        .into();

    let cluster: Cluster = serde_yaml::from_str(&cluster_v2_str).unwrap();

    let inventory_model_str: String = r#"---
all:
  vars:
    cartridge_app_name: myapp
    cartridge_cluster_cookie: myapp-cookie
    cartridge_package_path: /tmp/myapp.rpm
    cartridge_bootstrap_vshard: true
    cartridge_failover_params:
      mode: stateful
      state_provider: stateboard
      stateboard_params:
        uri: "192.168.16.11:4401"
        password: password
  hosts:
    storage-1:
      config:
        advertise_uri: "192.168.16.11:3031"
        http_port: 8081
    stateboard:
      stateboard: true
      config:
        listen: "192.168.16.11:4401"
        password: password
  children:
    storage-1-replicaset:
      vars:
        replicaset_alias: storage-1
        failover_priority:
          - storage-1
        roles:
          - storage
        all_rw: true
      hosts:
        storage-1: ~
    server-1:
      vars:
        ansible_host: 192.168.16.11
      hosts:
        storage-1: ~
        stateboard: ~"#
        .into();

    let inventory_model: Inventory = serde_yaml::from_str(&inventory_model_str).unwrap();

    let inventory = Inventory::try_from(&Some(cluster)).unwrap();

    assert_eq!(inventory, inventory_model);
}
