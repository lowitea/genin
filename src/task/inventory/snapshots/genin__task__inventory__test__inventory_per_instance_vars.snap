---
source: src/task/inventory/test.rs
expression: inventory
---
all:
  vars:
    ansible_user: ansible
    ansible_password: ansible
    cartridge_app_name: myapp
    cartridge_cluster_cookie: myapp-cookie
    cartridge_package_path: /tmp/myapp.rpm
    cartridge_bootstrap_vshard: true
    cartridge_failover_params:
      mode: stateful
      state_provider: stateboard
      stateboard_params:
        uri: "192.168.16.11:4401"
        password: password
  hosts:
    router-1:
      config:
        advertise_uri: "192.168.16.11:3031"
        http_port: 8081
      vars:
        cartridge_extra_env:
          TARANTOOL_NET_MSG_MAX: 1536
        cartridge_force_leader_control_instance: false
    stateboard:
      stateboard: true
      config:
        listen: "192.168.16.11:4401"
        password: password
  children:
    router-1-replicaset:
      vars:
        replicaset_alias: router-1
        failover_priority:
          - router-1
        roles:
          - router
      hosts:
        router-1: ~
    server-1:
      vars:
        ansible_host: 192.168.16.11
      hosts:
        router-1: ~
        stateboard: ~

